awx/api/views/instance_install_bundle.py

```diff
@@ -31,24 +55,181 @@ def get(self, request, *args, **kwargs):
                 status=status.HTTP_400_BAD_REQUEST,
             )
 
-        # TODO: add actual data into the bundle
-        # create a named temporary file directory to store the content of the install bundle
-        with tempfile.TemporaryDirectory() as tmpdirname:
-            # create a empty file named "moc_content.txt" in the temporary directory
-            with open(os.path.join(tmpdirname, 'mock_content.txt'), 'w') as f:
-                f.write('mock content')
-
-            # create empty directory in temporary directory
-            os.mkdir(os.path.join(tmpdirname, 'mock_dir'))
-
-            # tar.gz and create a temporary file from the temporary directory
-            # the directory will be renamed and prefixed with the hostname of the instance
-            with tempfile.NamedTemporaryFile(suffix='.tar.gz') as tmpfile:
-                with tarfile.open(tmpfile.name, 'w:gz') as tar:
-                    tar.add(tmpdirname, arcname=f"{instance_obj.hostname}_install_bundle")
-
-                # read the temporary file and send it to the client
-                with open(tmpfile.name, 'rb') as f:
-                    response = HttpResponse(f.read(), status=status.HTTP_200_OK)
-                    response['Content-Disposition'] = f"attachment; filename={instance_obj.hostname}_install_bundle.tar.gz"
-                    return response
+        with io.BytesIO() as f:
+            with tarfile.open(fileobj=f, mode='w:gz') as tar:
+                # copy /etc/receptor/tls/ca/receptor-ca.crt to receptor/tls/ca in the tar file
+                tar.add(
+                    os.path.realpath('/etc/receptor/tls/ca/receptor-ca.crt'), arcname=f"{instance_obj.hostname}_install_bundle/receptor/tls/ca/receptor-ca.crt"
+                )
+
+                # copy /etc/receptor/signing/work-public-key.pem to receptor/work-public-key.pem
+                tar.add('/etc/receptor/signing/work-public-key.pem', arcname=f"{instance_obj.hostname}_install_bundle/receptor/work-public-key.pem")
+
+                # generate and write the receptor key to receptor/tls/receptor.key in the tar file
+                key, cert = generate_receptor_tls(instance_obj)
+
+                key_tarinfo = tarfile.TarInfo(f"{instance_obj.hostname}_install_bundle/receptor/tls/receptor.key")
+                key_tarinfo.size = len(key)
+                tar.addfile(key_tarinfo, io.BytesIO(key))
+
+                cert_tarinfo = tarfile.TarInfo(f"{instance_obj.hostname}_install_bundle/receptor/tls/receptor.crt")
+                cert_tarinfo.size = len(cert)
+                tar.addfile(cert_tarinfo, io.BytesIO(cert))
+
+                # generate and write install_receptor.yml to the tar file
+                playbook = generate_playbook()
+                playbook_tarinfo = tarfile.TarInfo(f"{instance_obj.hostname}_install_bundle/install_receptor.yml")
+                playbook_tarinfo.size = len(playbook)
```

Comment:
This (and the similar patterns below) might cause problems in the presence of actual unicode characters, though at the moment I don't see that any can leak into these.  My recommendation would still be to do the `.encode('utf-8')` up front and pass the results to both `len()` and `io.BytesIO()`.

Reply:
fixed